src/librustdoc/clean/mod.rs

```diff
@@ -1698,11 +1717,197 @@ fn normalize<'tcx>(
     }
 }
 
+// FIXME(fmease): IIUC, the rules differ if we are in an “expression context” (see the Reference).
+// If so, add an extra parameter specifying this information. “Expression context” means if we are
+// inside of a function header or body, right?
+// FIXME(fmease): Check if we handle higher-rank lifetime bounds of various forms and combinations correctly.
+// I suspect our current naive lexical substitutions violate a lot of rules and are basically incorrect!
+fn clean_trait_object_lifetime_bound<'tcx>(
+    region: ty::Region<'tcx>,
+    container: Option<ContainerTy<'tcx>>,
+    trait_: DefId,
+    tcx: TyCtxt<'tcx>,
+    substs: &ty::List<ty::GenericArg<'tcx>>,
+) -> Option<Lifetime> {
+    // Below, we quote extracts from https://doc.rust-lang.org/reference/lifetime-elision.html#default-trait-object-lifetimes
+
+    // > If the trait object is used as a type argument of a generic type then the containing type is
+    // > first used to try to infer a bound.
+    let default = container
+        .map_or(ObjectLifetimeDefault::Empty, |container| container.object_lifetime_default(tcx));
+
+    // > If there is a unique bound from the containing type then that is the default
+    // If there is a default object lifetime and the given region is lexically equal to it, elide it.
+    let lifetime = match *region {
```

Comment:
I suggest doing it in the opposite order:
- compute what the implicit lifetime is from `ObjectLifetimeDefault`,
- compare it with `region`.